From aafcd378fd45a51298bd1a1bc892aeee6ac50a29 Mon Sep 17 00:00:00 2001
From: Varun Sharma <vsharma@habana.ai>
Date: Sun, 27 Feb 2022 16:42:49 +0200
Subject: [PATCH] [SW-43475] greco: enable decompression support for preboot arc

This patch enable support to decompress images like dsv which
are compressed via lzo to be decompressed in preboot running
on arc cpu. Convert fit 64 bit address to arc view point address,
as greco preboot run on arc cpu.

Change-Id: I140b12ec854a613419d0dd7af9aa351ff5d57eb4
Signed-off-by: Varun Sharma <vsharma@habana.ai>
---

diff --git a/arch/greco/arc/common.c b/arch/greco/arc/common.c
index bf642e6..c4084f0 100644
--- a/arch/greco/arc/common.c
+++ b/arch/greco/arc/common.c
@@ -61,13 +61,6 @@
 	memcpy((void *)arc_dst, buf, len);
 }
 
-int preboot_decompress_image(u64 dst, const void *buf, int len,
-			     const char *arch, const char *compression)
-{
-	preboot_puts("Compression is not supported\n");
-	return 1;
-}
-
 void preboot_start_image(u64 startaddr, bool valid_arc_img)
 {
 	void (*image_entry)(void);
@@ -219,3 +212,8 @@
 		}
 	}
 }
+
+u64 preboot_get_arch_view_addr(u64 addr)
+{
+	return ((u32)addr & 0xFFFFFFF) | BIT(28);
+}
diff --git a/lib/preboot/fit.c b/lib/preboot/fit.c
index e067c95..96e3db3 100644
--- a/lib/preboot/fit.c
+++ b/lib/preboot/fit.c
@@ -88,10 +88,15 @@
 	memcpy((void *)(uintptr_t)dst, buf, len);
 }
 
+__weak u64 preboot_get_arch_view_addr(u64 addr)
+{
+	return addr;
+}
+
 __weak int preboot_decompress_image(u64 dst, const void *buf, int len,
 				    const char *arch, const char *cmpr)
 {
-	void *dstptr = (void *)(uintptr_t)dst;
+	void *dstptr = (void *)(uintptr_t)preboot_get_arch_view_addr(dst);
 	size_t dst_len = SIZE_MAX;
 	int err;
 
